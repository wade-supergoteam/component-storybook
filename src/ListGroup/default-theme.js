// @flow

export const primary = "#007bff";
export const secondary = "#6c757d";
export const success = "#28a745";
export const info = "#17a2b8";
export const warning = "#ffc107";
export const danger = "#dc3545";
export const light = "#f8f9fa";
export const dark = "#343a40";

export const listGroupBg = "#fff";
export const listGroupBorderColor = "rgba(0,0,0,0.125)";
export const listGroupBorderWidth = "1px";
export const listGroupBorderRadius = ".25rem";
export const listGroupItemPaddingY = ".75rem";
export const listGroupItemPaddingX = "1.25rem";
export const listGroupHoverBg = "#f8f9fa";
export const listGroupActiveColor = "#fff";
export const listGroupActiveBg = "#007bff";
export const listGroupActiveBorderColor = "#007bff";
export const listGroupDisabledColor = "#6c757d";
export const listGroupDisabledBg = "#fff";
export const listGroupActionColor = "#495057";
export const listGroupActionHoverColor = "#495057";
export const listGroupActionActiveColor = "#212529";
export const listGroupActionActiveBg = "#e9ecef";
