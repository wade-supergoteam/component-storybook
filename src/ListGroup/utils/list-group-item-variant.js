// @ flow

import { css } from "styled-components";

function listGroupItemVariant(state: string, background: string, color: string) {
  return css`
    color: ${color};
    background-color: ${background};
  `;
}

export default listGroupItemVariant;
