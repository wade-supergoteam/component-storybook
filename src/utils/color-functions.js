// @flow

// $FlowIssue
import mix from "polished/lib/color/mix";
// $FlowIssue
import parseToRgb from "polished/lib/color/parseToRgb";

const themeColorInterval = 0.08;

// Color contrast
export function colorYiq(color: string): string {
  const r = parseToRgb(color).red;
  const g = parseToRgb(color).green;
  const b = parseToRgb(color).blue;

  const yiq = (r * 299 + g * 587 + b * 114) / 1000;

  if (yiq >= 187) {
    return "#111";
  }

  return "#fff";
}

// Request a theme color level
export function themeColorLevel(color: string, level: number = 0) {
  const colorBase = level > 0 ? "#111" : "#fff";

  if (level < 0) {
    // Lighter values need a quick double negative for the Sass math to work
    return mix(level * -1 * themeColorInterval, colorBase, color);
  }

  return mix(level * themeColorInterval, colorBase, color);
}
