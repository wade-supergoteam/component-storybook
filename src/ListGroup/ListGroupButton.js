// @flow

import styled, { css } from "styled-components";

// $FlowIssue
import get from "lodash.get";
// $FlowIssue
import darken from "polished/lib/color/darken";

import { hoverFocus } from "../utils/hover";
import { themeColorLevel } from "../utils/color-functions";
import themeProp from "../utils/theme";
import * as defaultTheme from "./default-theme";

import InputGroupItem from "./ListGroupItem";

const {
  listGroupHoverBg,
  listGroupDisabledColor,
  listGroupDisabledBg,
  listGroupActionColor,
  listGroupActionHoverColor,
  listGroupActionActiveColor,
  listGroupActiveColor,
  listGroupActionActiveBg,
  listGroupActiveBg,
} = defaultTheme;

// TODO: it has a .withComponent before
const ListGroupButton = styled(InputGroupItem)`
  width: 100%;
  color: ${themeProp("listGroupActionColor", listGroupActionColor)};
  text-align: inherit;

  ${({ active, theme }) =>
    active &&
    css`
      color: ${get(theme, "listGroupActiveColor", listGroupActiveColor)}!important;
      background-color: ${get(theme, "listGroupActiveBg", listGroupActiveBg)}!important;
    `}

  &:active {
    color: ${themeProp("listGroupActionActiveColor", listGroupActionActiveColor)};
    background-color: ${themeProp("listGroupActionActiveBg", listGroupActionActiveBg)};
  }

  /* Hover state */
  ${hoverFocus(css`
    color: ${themeProp("listGroupActionHoverColor", listGroupActionHoverColor)};
    text-decoration: none;
    background-color: ${themeProp("listGroupHoverBg", listGroupHoverBg)};
  `)}

  ${({ disabled, theme }) =>
    disabled &&
    css`
      color: ${get(theme, "listGroupDisabledColor", listGroupDisabledColor)};
      background-color: ${get(theme, "listGroupDisabledBg", listGroupDisabledBg)};
    `}

  ${({ color, theme }) => {
    if (color) {
      const colorX = themeColorLevel(get(theme, `${color}`, defaultTheme[color]), 6);

      const backgroundX = themeColorLevel(get(theme, `${color}`, defaultTheme[color]), -9);

      return css`
        ${hoverFocus(css`
          color: ${colorX};
          background-color: ${darken(0.05, backgroundX)};
        `)};

        &.active {
          color: #fff;
          background-color: ${colorX};
          border-color: ${colorX}};
        }
      `;
    }
  }};
`;

// ListGroupButton.defaultProps = {};

export default ListGroupButton;
