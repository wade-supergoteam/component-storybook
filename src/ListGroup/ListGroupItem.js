// @flow

import styled, { css } from "styled-components";
// $FlowIssue
import { borderRadius } from "polished";
// $FlowIssue
import get from "lodash.get";

import { hoverFocus } from "../utils/hover";
import listGroupItemVariant from "./utils/list-group-item-variant";
import themeProp from "../utils/theme";
import { themeColorLevel } from "../utils/color-functions";

import * as defaultTheme from "./default-theme";

const {
  listGroupBg,
  listGroupBorderColor,
  listGroupBorderWidth,
  listGroupBorderRadius,
  listGroupItemPaddingY,
  listGroupItemPaddingX,
  listGroupActiveColor,
  listGroupActiveBg,
  listGroupActiveBorderColor,
  listGroupDisabledColor,
  listGroupDisabledBg,
} = defaultTheme;

const ListGroupItem = styled.li`
  position: relative;
  display: block;
  padding: ${themeProp("listGroupItemPaddingY", listGroupItemPaddingY)}
    ${themeProp("listGroupItemPaddingX", listGroupItemPaddingX)};

  margin-bottom: -${themeProp("listGroupBorderWidth", listGroupBorderWidth)};
  background-color: ${themeProp("listGroupBg", listGroupBg)};

  border: ${themeProp("listGroupBorderWidth", listGroupBorderWidth)} solid
    ${themeProp("listGroupBorderColor", listGroupBorderColor)};

  &:first-child {
    ${borderRadius("top", listGroupBorderRadius)};
  }

  &:last-child {
    margin-bottom: 0;
    ${borderRadius("bottom", listGroupBorderRadius)};
  }

  ${hoverFocus(css`
    z-index: 1;
    text-decoration: none;
  `)};

  ${({ disabled, theme }) =>
    disabled &&
    css`
      color: ${get(theme, "listGroupDisabledColor", listGroupDisabledColor)};
      background-color: ${get(theme, "listGroupDisabledBg", listGroupDisabledBg)};
    `};

  ${({ active, theme }) =>
    active &&
    css`
      z-index: 2;
      color: ${get(theme, "listGroupActiveColor", listGroupActiveColor)};
      background-color: ${get(theme, "listGroupActiveBg", listGroupActiveBg)};
      border-color: ${get(theme, "listGroupActiveBorderColor", listGroupActiveBorderColor)};
    `};

  ${({ color, theme }) =>
    color &&
    listGroupItemVariant(
      get(theme, `${color}`, defaultTheme[color]),
      themeColorLevel(get(theme, `${color}`, defaultTheme[color]), -9),
      themeColorLevel(get(theme, `${color}`, defaultTheme[color]), 6),
    )};
`;

// ListGroupItem.defaultProps = {};

export default ListGroupItem;
