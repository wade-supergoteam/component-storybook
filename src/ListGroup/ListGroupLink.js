// @flow

import ListGroupButton from "./ListGroupButton";

const ListGroupLink = ListGroupButton.withComponent("a");

export default ListGroupLink;
