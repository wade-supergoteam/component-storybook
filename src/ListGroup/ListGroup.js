// @flow

import styled, { css } from "styled-components";

import ListGroupItem from "./ListGroupItem";
import ListGroupButton from "./ListGroupButton";
import ListGroupLink from "./ListGroupLink";

const ListGroup = styled.ul`
  display: flex;
  flex-direction: column;
  padding-left: 0;
  margin-bottom: 0;

  ${({ flush }) =>
    flush &&
    css`
      ${ListGroupItem} {
        border-right: 0;
        border-left: 0;
        border-radius: 0;
      }

      &:first-child {
        ${ListGroupItem}:first-child {
          border-top: 0;
        }
      }

      ${ListGroupItem}:last-child {
        border-bottom: 0 !important;
      }
    `};
`;

ListGroup.Item = ListGroupItem;
ListGroup.Button = ListGroupButton;
ListGroup.Link = ListGroupLink;

export default ListGroup;
