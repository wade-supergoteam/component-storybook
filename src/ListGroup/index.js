import ListGroup from "./ListGroup";

import ListGroupItem from "./ListGroupItem";
import ListGroupButton from "./ListGroupButton";
import ListGroupLink from "./ListGroupLink";

export { ListGroupItem, ListGroupButton, ListGroupLink };
export default ListGroup;
