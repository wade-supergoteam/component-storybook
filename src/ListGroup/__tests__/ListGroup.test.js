import "jest-styled-components";
import React from "react";
import renderer from "react-test-renderer";
import ListGroup from "../";
import ListGroupButton from "../ListGroupButton";
import ListGroupItem from "../ListGroupItem";
import ListGroupLink from "../ListGroupLink";
import { themeColorLevel } from "../../utils/color-functions";

test("Basic Example", () => {
  const wrapper = renderer.create(<ListGroup />);
  expect(wrapper).toMatchSnapshot();
});

test("ListGroupButton active", () => {
  const wrapper = renderer.create(<ListGroupButton active />);
  expect(wrapper).toMatchSnapshot();
});

test("ListGroupButton disabled", () => {
  const wrapper = renderer.create(<ListGroupButton disabled />);
  expect(wrapper).toMatchSnapshot();
});

test("ListGroupButton color", () => {
  const wrapper = renderer.create(<ListGroupButton color="primary" />);
  expect(wrapper).toMatchSnapshot();
});

test("ListGroupButton", () => {
  const wrapper = renderer.create(<ListGroupButton />).toJSON();
  expect(wrapper).toHaveStyleRule("color", "#495057");
  expect(wrapper).toHaveStyleRule("background-color", "#fff");
});

test("Flush List group", () => {
  const wrapper = renderer.create(<ListGroup flush />);
  expect(wrapper).toMatchSnapshot();
});

test("Disabled items", () => {
  const wrapper = renderer.create(<ListGroupItem disabled />);
  expect(wrapper).toMatchSnapshot();
});

test("Links", () => {
  const wrapper = renderer.create(<ListGroupLink href="#" />);
  expect(wrapper).toMatchSnapshot();
});
