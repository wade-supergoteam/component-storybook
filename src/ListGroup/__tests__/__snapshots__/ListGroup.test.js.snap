// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Example 1`] = `
<ul
  className="sc-bxivhb gFoUmm"
/>
`;

exports[`Buttons 1`] = `
<li
  className="sc-bdVaJa sc-bwzfXH ocRaz"
/>
`;

exports[`Disabled items 1`] = `
<li
  className="sc-bdVaJa kQsWJh"
  disabled={true}
/>
`;

exports[`Flush List group 1`] = `
<ul
  className="sc-bxivhb huWLJC"
/>
`;

exports[`Links 1`] = `
<a
  className="sc-htpNat dXdgFK"
  href="#"
/>
`;
